src/Umbraco.Infrastructure/Persistence/Relations/ContentRelationsUpdate.cs

```diff
@@ -0,0 +1,159 @@
+using Microsoft.Extensions.Logging;
+using NPoco;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.Cache;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Models.Editors;
+using Umbraco.Cms.Core.Notifications;
+using Umbraco.Cms.Core.Persistence.Querying;
+using Umbraco.Cms.Core.Persistence.Repositories;
+using Umbraco.Cms.Core.PropertyEditors;
+using Umbraco.Cms.Infrastructure.Persistence.Dtos;
+using Umbraco.Cms.Infrastructure.Scoping;
+using Umbraco.Extensions;
+
+namespace Umbraco.Cms.Infrastructure.Persistence.Relations;
+
+/// <summary>
+/// Defines a notification handler for content saved operations that persists relations.
+/// </summary>
+internal class ContentRelationsUpdate :
+    IDistributedCacheNotificationHandler<ContentSavedNotification>,
+    IDistributedCacheNotificationHandler<ContentPublishedNotification>,
+    IDistributedCacheNotificationHandler<MediaSavedNotification>,
+    IDistributedCacheNotificationHandler<MemberSavedNotification>
+{
+    private readonly IScopeProvider _scopeProvider;
+    private readonly DataValueReferenceFactoryCollection _dataValueReferenceFactories;
+    private readonly PropertyEditorCollection _propertyEditors;
+    private readonly IRelationRepository _relationRepository;
+    private readonly IRelationTypeRepository _relationTypeRepository;
+    private readonly ILogger<ContentRelationsUpdate> _logger;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ContentRelationsUpdate"/> class.
+    /// </summary>
+    public ContentRelationsUpdate(
+        IScopeProvider scopeProvider,
+        DataValueReferenceFactoryCollection dataValueReferenceFactories,
+        PropertyEditorCollection propertyEditors,
+        IRelationRepository relationRepository,
+        IRelationTypeRepository relationTypeRepository,
+        ILogger<ContentRelationsUpdate> logger)
+    {
+        _scopeProvider = scopeProvider;
+        _dataValueReferenceFactories = dataValueReferenceFactories;
+        _propertyEditors = propertyEditors;
+        _relationRepository = relationRepository;
+        _relationTypeRepository = relationTypeRepository;
+        _logger = logger;
+    }
+
+    /// <inheritdoc/>
+    public void Handle(ContentSavedNotification notification) => PersistRelations(notification.SavedEntities);
+
+    /// <inheritdoc/>
+    public void Handle(ContentPublishedNotification notification) => PersistRelations(notification.PublishedEntities);
+
+    /// <inheritdoc/>
+    public void Handle(MediaSavedNotification notification) => PersistRelations(notification.SavedEntities);
+
+    /// <inheritdoc/>
+    public void Handle(MemberSavedNotification notification) => PersistRelations(notification.SavedEntities);
```

Comment:
We should also override the `void Handle(IEnumerable<TNotification> notifications)` overload to avoid creating separate scopes when multiple notifications are handled in batches (see https://github.com/umbraco/Umbraco-CMS/pull/14332).